### PR TITLE
add millisecond and microsecond format strings

### PR DESCRIPTION
**Pull request by rgarcia**
*Issue #, if available:* https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/issues/34

*Description of changes:*

It's important when looking at time-ordered logs to have resolution
below a second. I think this benefit outweighs being strictly
POSIX-compliant. This change allows for the format specifiers `%L` for
milliseconds and `%f` for microseconds, which are the format strings
used in the extensions provided by the github.com/lestrrat-go/strftime
library.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by hencrice** [comment]
I don't see a downside for doing this. Any reason why we chose to stick to POSIX-compliant initially? @PettitWesley ?

**Comment by PettitWesley** [comment]
@hencrice @rgarcia One of my goals was to make it easy to re-write this plugin in C and contribute it to the core of Fluent Bit in the future. Are those format specifiers present in Posix strftime? I'm not sure they are: https://www.man7.org/linux/man-pages/man3/strftime.3.html

That was my intent... I am fine with this change I guess... if we do decide to rewrite this plugin in C, I think we could build a shim around strftime that adds support for those format specifiers (otherwise we could never deprecate this version).

**Comment by rgarcia** [comment]
@PettitWesley my read of the situation from googling is that strict POSIX strftime doesn't have sub-second resolution, which means over the years everyone has filled in that gap in their own way, without a standard emerging. Python uses `%L` and `%f` for millis and nanos (like this PR). The [boost](https://www.boost.org/doc/libs/1_35_0/doc/html/date_time/date_time_io.html#date_time.format_flags) C++ library has `%s` for fractional seconds e.g. (59.000000). Golang has their own quirky time formatting but it supports nanos ([example](https://play.golang.org/p/T_syDPGur8F). Ruby has `%L` for millis and `%N` for nanos. 

From our perspective it's hard to use the library as is, since when we analyze logs it's important to have timestamps that are sub-second. So we'd appreciate having this feature.

**Comment by PettitWesley** [comment]
@rgarcia I see, so this follows Python conventions. I guess that's alright with me.

**Review by sonofachamp — APPROVED** [review]

**Comment by rgarcia** [comment]
FWIW when I run this in ECS I see timestamps that have `000000` for microseconds, suggesting that fluent bit is passing a unix timestamp to the plugin: https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/blob/09736e266f272754473087b346c0c60ebd35a069/fluent-bit-kinesis.go#L136

Not familiar with how the fluent bit plugin system works so could use an assist in figuring out what is happening here...

**Comment by PettitWesley** [comment]
@rgarcia Check this out- looks like Fluent Bit supports nanoseconds and uses %L (on the input side for parsing timestamps): https://fluentbit.io/documentation/0.13/parser/

**Comment by rgarcia** [comment]
@PettitWesley still not sure what's going on... our application logs do not contain timestamps, so we are dependent on the timestamps created by the system receiving these logs (in the past: syslog, future: firelens). Is there some fluent-bit configuration that controls the timestamp passed to plugins in the case where none can be parsed from the application log?

I played around with commenting out this block https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/blob/09736e266f272754473087b346c0c60ebd35a069/fluent-bit-kinesis.go#L133-L136 
and it turns out that it was not that case, but the case above that was getting hit:
https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/blob/09736e266f272754473087b346c0c60ebd35a069/fluent-bit-kinesis.go#L131-L132
So something is causing `FLBTime` to only have seconds resolution.

I went ahead and commented out everything so that the default case always gets hit: https://github.com/aws/amazon-kinesis-streams-for-fluent-bit/blob/09736e266f272754473087b346c0c60ebd35a069/fluent-bit-kinesis.go#L137-L138

and I finally saw sub-second resolution, but strangely only timestamps of the form HH:MM:SS.000XYZ--i.e. tenths, hundredths, and thousandths of a second were always 0. Very confused by this!

**Comment by PettitWesley** [comment]
@rgarcia Read through this header file: https://github.com/fluent/fluent-bit/blob/master/include/fluent-bit/flb_time.h

FLB_TIME is a wrapper around timespec, which has nanoseconds. But, if you go through that code, depending on the CMake flags, only seconds may be used and nanoseconds will always be set to zero. I'm guessing that is the problem...

Let me reach out to the lead maintainer of Fluent Bit and ask him about this.

In the meantime, let's hold off on merging this feature until we have a full solution. We may possibly want to introduce an additional parameter to auto-create new timestamps (with full resolution).

Edit: sorry, meant to reference the code here: https://github.com/fluent/fluent-bit/blob/master/src/flb_time.c#L48

**Review by PettitWesley — CHANGES_REQUESTED** [review]
Blocking merge until we discuss with the Fluent Bit lead maintainer.

**Comment by PettitWesley** [comment]
> so we are dependent on the timestamps created by the system receiving these logs (in the past: syslog, future: firelens)

@rgarcia So in FireLens, the timestamps from the Fluentd Docker Log Driver- from what I can tell, Fluent Bit does not add a timestamp, it just copies what it gets from Docker.

I tried to dig through the docker code a bit... it seems the timestamp is set in the log message here: https://github.com/moby/moby/blob/master/daemon/logger/adapter.go#L113

That code really makes it look like it supports nanoseconds. But in my experience i think I've only ever seen second resolution timestamps...

**Comment by rgarcia** [comment]
Thanks @PettitWesley. I've tried compiling fluent-bit with `-DFLB_CLOCK_GET_TIME=On` (https://github.com/aws/aws-for-fluent-bit/blob/69b5974f20748e387fdbdaaafcdae4feb64699c1/Dockerfile#L42), but the `FLBTime` received in the kinesis plugin is still only seconds resolution, and `time.Now()` still has `000` for tenths, hundredths, and thousandths of a second (example output of [this](https://github.com/Clever/amazon-kinesis-streams-for-fluent-bit/blob/117b25b759f9fb4da5a5e604b9de97a169dd1f6c/fluent-bit-kinesis.go#L135) `fmt.Println` is `DEBUG FLBTime timestamp is  2020-07-12T12:28:23Z 2020-07-12T12:28:27.00087163Z`) . I'd like to try modifying the aws-for-fluent-bit build process to turn off all fluent-bit parsing of the log generated by the docker daemon, since it seems like this might be the source of the issue, but it's not clear to me how to do this. Any assist would be appreciated.

I also still don't fully understand how `time.Now()` could always return these leading `000`s in the kinesis plugin, and how build flags for fluent affect this.

**Review by PettitWesley — APPROVED** [review]

**Comment by PettitWesley** [comment]
@rgarcia I talked to the Fluent Bit lead maintainer, unfortunately, he does not have experience with the Fluentd docker log driver.

However, I am nearly certain that Fluent Bit is not the source of the problem here. In other modes, it can handle timestamps with millisecond resolution. I am pretty sure that the Fluentd Docker log driver is sending the timestamps with only second resolution.

To try to validate this, I send logs to Fluentd with the docker log driver, and the same thing happened, they only had second resolution.

I will merge this PR; this feature is complete. To fully solve your use case, I think we need to also add an option in the plugin to generate new timestamps. @rgarcia Can you do that?

> tenths, hundredths, and thousandths of a second were always 0.

I have no clue about this... may be its a side-effect of something in the go runtime or implementation of time.Now()?

**Comment by rgarcia** [comment]
@PettitWesley I suspect the strange values for `time.Now()` are a result of fact that Go plugins receive a flush of data on an interval. `FLBPluginFlushCtx` gets called according to the interval passed via `-f, --flush=SECONDS    flush timeout in seconds (default: 5)`. It looks like the firelens image sets this to 1: https://github.com/aws/aws-for-fluent-bit/blob/55f5eb78bb8441a455cdcb0e1e32e92896c1e2a6/configs/minimize-log-loss.conf#L2

So even if this plugin had an option to generate its own timestamps, they would be incorrect :-/. I think the only option is to figure out how to get fluent-bit itself to generate a new timestamp regardless of what it parses from the log line.

**Comment by rgarcia** [comment]
Upon further investigation I think this is the issue we're running into: https://github.com/aws/containers-roadmap/issues/839 (found via https://github.com/fluent/fluent-bit/issues/572). For some reason the `fluentd` docker log driver defaults to seconds resolution (the `fluentd-sub-second-precision` option is set to `false`: https://docs.docker.com/config/containers/logging/fluentd/). Firelens does not change this default. This makes me think there's no real workaround here until the `firelensConfiguration` field in the task definition allows for setting this value to true.

**Comment by PettitWesley** [comment]
@rgarcia Thanks for that. I will work on making sure the container roadmap issue you linked gets prioritized.

**Pull request merged**
